### PR TITLE
build: update patch with bug fix to runtime grunning

### LINKED ISSUE #95529: grunning: `grunning.Time()` is non-monotonic
**Describe the problem**

In https://github.com/cockroachdb/cockroach/pull/93952 we observed non-monotonicity in `grunning.Time()`. The problem is due to an accounting bug in the [patched runtime](https://github.com/cockroachdb/cockroach/blob/f27fa9d/docs/RFCS/20220602_fine_grained_cpu_attribution.md#design) we're now running with. To maintain the nanoseconds spent per-g in the `_Grunning` state, we want to intercept all transitions in and out of the `_Grunning` state. 

1. When we transition into the `_Grunning` state, we maintain a timestamp on `type g struct` to record when we were scheduled:

https://github.com/cockroachdb/cockroach/blob/f27fa9d4c56e4654939e772d857027da00feb7d5/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L11-L16

2. When we transition out of the `Grunning` state, we accumulate the time we ran in the most recent interval:

https://github.com/cockroachdb/cockroach/blob/f27fa9d4c56e4654939e772d857027da00feb7d5/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L18-L21

As for surfacing how much time the calling goroutine has spent running, we use the last scheduled timestamp, accumulated counter when last descheduled, and a current monotonic clock reading, to get to the following:

https://github.com/cockroachdb/cockroach/blob/f27fa9d4c56e4654939e772d857027da00feb7d5/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L45-L51

This works, except that the patch does not actually intercept all transitions in and out of the `_Grunning` state. Specifically, we're missing the following points in the Go scheduler code:

```diff
diff --git i/src/runtime/proc.go w/src/runtime/proc.go
diff --git i/src/runtime/proc.go w/src/runtime/proc.go
index 4f1bed23d2..76871ad52e 100644
--- i/src/runtime/proc.go
+++ w/src/runtime/proc.go
@@ -929,6 +929,9 @@ func casfrom_Gscanstatus(gp *g, oldval, newval uint32) {
                dumpgstatus(gp)
                throw("casfrom_Gscanstatus: gp->status is not in scan state")
        }
+       if newval == _Grunning {
+               gp.lastsched = nanotime()
+       }
        releaseLockRank(lockRankGscan)
 }

@@ -944,6 +947,9 @@ func castogscanstatus(gp *g, oldval, newval uint32) bool {
                        r := atomic.Cas(&gp.atomicstatus, oldval, newval)
                        if r {
                                acquireLockRank(lockRankGscan)
+                               if oldval == _Grunning {
+                                       gp.runningnanos += nanotime() - gp.lastsched
+                               }
                        }
                        return r

@@ -1065,6 +1071,7 @@ func casGToPreemptScan(gp *g, old, new uint32) {
        }
        acquireLockRank(lockRankGscan)
        for !atomic.Cas(&gp.atomicstatus, _Grunning, _Gscan|_Gpreempted) {
+               gp.runningnanos += nanotime() - gp.lastsched
        }
 }
```

With it, the reproduction below no longer reproduces the regression.

**To Reproduce**

With that branch, and the following assertion of monotonicity, it reliably fails after a few seconds when using `dev test --stress  pkg/sql/opt/exec/explain -f TestCPUTimeEndToEnd`:

```diff
--- a/pkg/sql/colflow/stats.go
+++ b/pkg/sql/colflow/stats.go
@@ -23,6 +23,7 @@ import (
        "github.com/cockroachdb/cockroach/pkg/sql/execstats"
        "github.com/cockroachdb/cockroach/pkg/util/buildutil"
        "github.com/cockroachdb/cockroach/pkg/util/grunning"
+       "github.com/cockroachdb/cockroach/pkg/util/log"
        "github.com/cockroachdb/cockroach/pkg/util/mon"
        "github.com/cockroachdb/cockroach/pkg/util/syncutil"
        "github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -123,7 +124,14 @@ func (bic *batchInfoCollector) Next() coldata.Batch {
        bic.stopwatch.Start()
        // Wrap the call to Next() with a panic catcher in order to get the correct
        // execution time (e.g. in the statement bundle).
+       startCPUTime := grunning.Time()
        err := colexecerror.CatchVectorizedRuntimeError(bic.next)
+       endCPUTime := grunning.Time()
+       if endCPUTime < startCPUTime {
+               log.Fatalf(context.TODO(), "non-monotonicity! endCPUTime (%d) < startCPUTime (%d) ",
+                       endCPUTime, startCPUTime,
+               )
+       }
        bic.stopwatch.Stop()
        if err != nil {
                colexecerror.InternalError(err)
```

**Impact**

Before updating the patch, and all the testing/{micro,macro}benchmarking it entails, lets assess the impact. Since we're missing interception points in and out of `_Grunning`, what could go wrong? Given `grunning.Time()` is powered by a `grunningnanos()`, which is `gp.runningnanos + nanotime() - gp.lastsched`, we can break this down to the following:

- `gp.runningnanos` is maintained when leaving `_Grunning`. So it might be older/smaller than it actually is. So `grunningnanos()` can be an under estimate, and thus non-monotonic.

- `gp.lastsched` is maintained when entering `_Grunning`. So it might be older/smaller than it actually is. So `grunningnanos()` can be an over-estimate. It might also lead to `gp.runningnanos` being higher than it's supposed to be (since we use `gp.runningnanos += nanotime() - gp.lastsched` to keep it up-to-date), so `grunningnanos()` is again an over-estimate.

If the analysis above holds, given our current uses of `grunning.Time()`, we could paper over the non-monotonicity (zero out differences between end/start time if the latter is smaller than the former), and document its inaccuracy.

+cc @DrewKimball, @kvoli, @sumeerbhola.

Jira issue: CRDB-23571

Epic CRDB-34249

### PR DESCRIPTION
**Pull request by aadityasondhi**
In our previous version of this patch, we missed two entry points and one exit point into the grunning state of a Goroutine. This led to `grunning.Time()` being non-monotonic.

This new patch adds those missing integration points.

Fixes #95529.

Release note: None

### CODE REVIEW

**Review by rickystewart — APPROVED** [review]
Stamp

**Comment by rickystewart** [comment]
Started [TeamCity build](https://teamcity.cockroachdb.com/buildConfiguration/Internal_Cockroach_Build_Toolchains_BuildAndPublishPatchedGo/13894498?hideProblemsFromDependencies=false&hideTestsFromDependencies=false)

**Comment by aadityasondhi** [comment]
Previous patch was corrupted because I tried to minimize the lines changed. This one applies well but changes a lot more lines. Am I doing it correctly @rickystewart?

Steps I followed:
1. I checkout out the `go1.21.3` branch.
2. Applied the old patches
3. Made my changes
4. Copied the output of `git diff` into the patch file

**Comment by rickystewart** [comment]
Can you put the diff for `src/context/context.go` back on the bottom of the file? I know it's not in alphabetical order, but that's how it was for whatever reason. It'll be easier to see what you actually changed if the filenames are in the same order.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: regarding the actual code changes modulo the need for code comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118907#-:-Nq4J3VN69Jnj-r5sbN3:b-bcuzdk)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`build/teamcity/internal/release/build-and-publish-patched-go/diff.patch` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118907#-Nq4HoOIA_0FvyCUPCGj:-Nq4HoOIA_0FvyCUPCGk:b-yfbz0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/221e0e41612e269280f85a716711cf38f0e67176/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L113)):*
> ```Diff
> --- a/src/runtime/proc.go
> +++ b/src/runtime/proc.go
> @@ -1004,6 +1004,9 @@ func casfrom_Gscanstatus(gp *g, oldval, newval uint32) {
> ```

We have two comments in the existing patch:

```
// We're transitioning into the running state, record the timestamp for  
// subsequent use.
```

```
// We're transitioning out of running, record how long we were in the  
// state.
```

can you add the relevant comment to these new changes too.

File: build/teamcity/internal/release/build-and-publish-patched-go/diff.patch [hunk 1]
```diff
@@ -148,18 +176,20 @@ diff --git a/src/runtime/sizeof_test.go b/src/runtime/sizeof_test.go
...
+@@ -21,7 +21,7 @@ func TestSizeof(t *testing.T) {
  		_32bit uintptr // size on 32bit platforms
  		_64bit uintptr // size on 64bit platforms
  	}{
 -		{runtime.G{}, 252, 408},   // g, but exported for testing
 +		{runtime.G{}, 260, 416},   // g, but exported for testing
  		{runtime.Sudog{}, 56, 88}, // sudog, but exported for testing
  	}
-diff -ru a/src/context/context.go b/src/context/context.go
+
```
Comment (rickystewart) [inline build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:187]: Blank line here seems wrong

**Comment by rickystewart** [comment]
Trying another build, we'll see if [this](https://teamcity.cockroachdb.com/buildConfiguration/Internal_Cockroach_Build_Toolchains_BuildAndPublishPatchedGo/13899729?hideProblemsFromDependencies=false&hideTestsFromDependencies=false) works.

File: build/teamcity/internal/release/build-and-publish-patched-go/diff.patch (hunk 1 above)
Comment (rickystewart) [reply build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:187]: To be clear: if the diff apply works, don't worry about removing the blank line. Just a thought I had as I was reviewing.

**Comment by rickystewart** [comment]
Looks like it worked. I updated the PR and then we'll see if CI is happy.

**Comment by aadityasondhi** [comment]
Had to rebase on top of the changes made in https://github.com/cockroachdb/cockroach/pull/118605.

@rickystewart let me know if this is good to merge

**Review by aadityasondhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118907)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rickystewart and @sumeerbhola)

___
*[`build/teamcity/internal/release/build-and-publish-patched-go/diff.patch` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118907#-Nq4HoOIA_0FvyCUPCGj:-Nr6xhmRBP3FR3rIQivG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/221e0e41612e269280f85a716711cf38f0e67176/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L113)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We have two comments in the existing patch:

```
// We're transitioning into the running state, record the timestamp for  
// subsequent use.
```

```
// We're transitioning out of running, record how long we were in the  
// state.
```

can you add the relevant comment to these new changes too.
</blockquote></details>

Done.

**Review by aadityasondhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118907)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rickystewart and @sumeerbhola)

___
*[`build/teamcity/internal/release/build-and-publish-patched-go/diff.patch` line 187 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/118907#gh-1482246761:-Nr6y1G_EjX8aiEmuc4o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cf5e85d6fd030a7e76483e3ba0a1e2aff8505a5/build/teamcity/internal/release/build-and-publish-patched-go/diff.patch#L187)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

To be clear: if the diff apply works, don't worry about removing the blank line. Just a thought I had as I was reviewing.
</blockquote></details>

Done.

**Comment by rickystewart** [comment]
> @rickystewart let me know if this is good to merge

No. We need to build the SDK and incorporate it into your PR. I can get this process started for you.

**Comment by aadityasondhi** [comment]
> > @rickystewart let me know if this is good to merge
> 
> No. We need to build the SDK and incorporate it into your PR. I can get this process started for you.

Thanks! let me know if I need to do anything to move this forward.

**Comment by rickystewart** [comment]
@aadityasondhi Go ahead and apply this diff and we'll see if CI is okay with this.

```
diff --git a/WORKSPACE b/WORKSPACE
index dd157813027..31fddffb5b1 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -165,13 +165,13 @@ load(
 go_download_sdk(
     name = "go_sdk",
     sdks = {
-        "darwin_amd64": ("go1.21.5.darwin-amd64.tar.gz", "9a9af12fddf7315db68ceeb2980d9cecbf058d50356787bc1680417d9c5892eb"),
-        "darwin_arm64": ("go1.21.5.darwin-arm64.tar.gz", "7a5c1f1ca0831980ea06e7383a7e56af7665f8863bc923e19fd050c9fd4866a9"),
-        "linux_amd64": ("go1.21.5.linux-amd64.tar.gz", "7656c08487ea9d8817047be5698bf96a787c195e237d62c2d605ad754c55022d"),
-        "linux_arm64": ("go1.21.5.linux-arm64.tar.gz", "5ca30901c990538e21a8adf4a0945b013ecae34cfbfc3fbac893ac24c2c33c0c"),
-        "windows_amd64": ("go1.21.5.windows-amd64.tar.gz", "f32ad2e23031dc7e43db29dce7bf0cb1bba095760f96972ace094159f6bd2986"),
+        "darwin_amd64": ("go1.21.5.darwin-amd64.tar.gz", "182c3e597111d474b29d72fbfee1ddb2915acd39b7c3e534a78eaddfa767089d"),
+        "darwin_arm64": ("go1.21.5.darwin-arm64.tar.gz", "1894e6f86814cd882c51bb5a73cf1312551c734bd4ac9fb33866df2a50e1d1a0"),
+        "linux_amd64": ("go1.21.5.linux-amd64.tar.gz", "ff04d463bf9da03cfe6aefeff6039312c701f74ab291162b937b31026f262228"),
+        "linux_arm64": ("go1.21.5.linux-arm64.tar.gz", "9a09561064580e837af46bc314536b450bd2546985955ddf1e70699a4abe2fb6"),
+        "windows_amd64": ("go1.21.5.windows-amd64.tar.gz", "93375cd3e51b811297a52a0e7a9d98b126da3ae02daaabb5d423e1c496275f06"),
     },
-    urls = ["https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/{}"],
+    urls = ["https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/{}"],
     version = "1.21.5",
 )
 
@@ -659,8 +659,8 @@ go_download_sdk(
     # able to provide additional diagnostic information such as the expected version of OpenSSL.
     experiments = ["boringcrypto"],
     sdks = {
-        "linux_amd64": ("go1.21.5fips.linux-amd64.tar.gz", "33bc9129c5a9ebc3a5f6deb3e651f05216a035479404c849d443d0930aec2b6a"),
+        "linux_amd64": ("go1.21.5fips.linux-amd64.tar.gz", "68fefc15b4328244fe78f1269e8856a2a2cd52f1fb9f451bb8bc0c23d590fe14"),
     },
-    urls = ["https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/{}"],
+    urls = ["https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/{}"],
     version = "1.21.5fips",
 )
diff --git a/build/bazelutil/distdir_files.bzl b/build/bazelutil/distdir_files.bzl
index f7dbe5ad2bb..c2930425459 100644
--- a/build/bazelutil/distdir_files.bzl
+++ b/build/bazelutil/distdir_files.bzl
@@ -1202,12 +1202,12 @@ DISTDIR_FILES = {
     "https://storage.googleapis.com/public-bazel-artifacts/c-deps/20230718-202534/libproj_foreign.macos.20230718-202534.tar.gz": "96771a33542beb72067afcafaeb790134014e56798fa4cbe291894c4ebf8b68d",
     "https://storage.googleapis.com/public-bazel-artifacts/c-deps/20230718-202534/libproj_foreign.macosarm.20230718-202534.tar.gz": "b2c60ffe1f50c6e81ba906f773b95d3a6699538d57e71749579552f4211a1e3e",
     "https://storage.googleapis.com/public-bazel-artifacts/c-deps/20230718-202534/libproj_foreign.windows.20230718-202534.tar.gz": "16de1e76ee8de4bd144dc57bfde05385d086943ca1b64cc246055c8b0cd71c65",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5.darwin-amd64.tar.gz": "9a9af12fddf7315db68ceeb2980d9cecbf058d50356787bc1680417d9c5892eb",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5.darwin-arm64.tar.gz": "7a5c1f1ca0831980ea06e7383a7e56af7665f8863bc923e19fd050c9fd4866a9",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5.linux-amd64.tar.gz": "7656c08487ea9d8817047be5698bf96a787c195e237d62c2d605ad754c55022d",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5.linux-arm64.tar.gz": "5ca30901c990538e21a8adf4a0945b013ecae34cfbfc3fbac893ac24c2c33c0c",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5.windows-amd64.tar.gz": "f32ad2e23031dc7e43db29dce7bf0cb1bba095760f96972ace094159f6bd2986",
-    "https://storage.googleapis.com/public-bazel-artifacts/go/20240201-224012/go1.21.5fips.linux-amd64.tar.gz": "33bc9129c5a9ebc3a5f6deb3e651f05216a035479404c849d443d0930aec2b6a",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5.darwin-amd64.tar.gz": "182c3e597111d474b29d72fbfee1ddb2915acd39b7c3e534a78eaddfa767089d",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5.darwin-arm64.tar.gz": "1894e6f86814cd882c51bb5a73cf1312551c734bd4ac9fb33866df2a50e1d1a0",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5.linux-amd64.tar.gz": "ff04d463bf9da03cfe6aefeff6039312c701f74ab291162b937b31026f262228",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5.linux-arm64.tar.gz": "9a09561064580e837af46bc314536b450bd2546985955ddf1e70699a4abe2fb6",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5.windows-amd64.tar.gz": "93375cd3e51b811297a52a0e7a9d98b126da3ae02daaabb5d423e1c496275f06",
+    "https://storage.googleapis.com/public-bazel-artifacts/go/20240220-195058/go1.21.5fips.linux-amd64.tar.gz": "68fefc15b4328244fe78f1269e8856a2a2cd52f1fb9f451bb8bc0c23d590fe14",
     "https://storage.googleapis.com/public-bazel-artifacts/java/railroad/rr-1.63-java8.zip": "d2791cd7a44ea5be862f33f5a9b3d40aaad9858455828ebade7007ad7113fb41",
     "https://storage.googleapis.com/public-bazel-artifacts/js/rules_jest-v0.18.4.tar.gz": "d3bb833f74b8ad054e6bff5e41606ff10a62880cc99e4d480f4bdfa70add1ba7",
     "https://storage.googleapis.com/public-bazel-artifacts/js/rules_js-v1.26.1.tar.gz": "08061ba5e5e7f4b1074538323576dac819f9337a0c7d75aee43afc8ae7cb6e18",

```

**Comment by aadityasondhi** [comment]
> @aadityasondhi Go ahead and apply this diff and we'll see if CI is okay with this.
> 

Build is looking good, feel free to bors if you think it is good to do so.

**Comment by aadityasondhi** [comment]
bors r=sumeerbhola,rickystewart

**Pull request merged**